File: frontend/src/api/schemas.ts
/**
 * Zod schemas validating unknown JSON at the API/transport boundary (PPI-022/030/034).
 *
 * The frontend receives raw JSON from `fetch` (HTTP) or `postMessage` (webview).
 * Before any of it reaches selectors/components, it must be parsed here so a
 * malformed backend/bridge response is a typed `DecodeError`, not a silent
 * `undefined` crash later.
 */
import { z } from "zod";

export const EdgeBreakdownSchema = z.record(z.string(), z.number());

const readonlyStringRecord = z.record(z.string(), z.number());

export const UiOptionSchema = z.object({
  id: z.string(),
  label: z.string(),
  default_enabled: z.boolean().optional(),
});

export const UiMetricOptionSchema = z.object({
  id: z.string(),
  label: z.string(),
  unit: z.string().optional(),
  format: z.string().optional(),
  default_enabled: z.boolean().optional(),
});

export const UiColumnDefinitionSchema = z.object({
  key: z.string(),
  label: z.string(),
  type: z.string().optional(),
  metric_id: z.string().nullable().optional(),
  width: z.number().nullable().optional(),
});

export const UiTableDefinitionSchema = z.object({
  key: z.string(),
  label: z.string(),
  columns: z.array(UiColumnDefinitionSchema),
});

export const UiGraphConfigSchema = z.object({
  edge_types: z.array(UiOptionSchema),
  line_categories: z.array(UiOptionSchema),
  brightness_metrics: z.array(UiMetricOptionSchema),
  node_size_metrics: z.array(UiMetricOptionSchema),
  link_thickness_metrics: z.array(UiMetricOptionSchema),
});

export const UiConfigResponseSchema = z.object({
  dashboard_metrics: z.array(UiMetricOptionSchema),
  aggregations: z.array(UiOptionSchema),
  tables: z.array(UiTableDefinitionSchema),
  graph: UiGraphConfigSchema,
});

export const GenericTableRowSchema = z.object({
  id: z.string().optional(),
  cells: z.record(z.string(), z.unknown()),
  actions: z.record(z.string(), z.boolean()).optional(),
});

export const GenericTableResponseSchema = z.object({
  commit_hash: z.string(),
  rows: z.array(GenericTableRowSchema),
});

export const RelationRowSchema = z.object({
  source_id: z.string(),
  source_label: z.string(),
  target_id: z.string(),
  target_label: z.string(),
  relation_type_id: z.string(),
  relation_type_label: z.string(),
  strength_metric_id: z.string().optional(),
  strength_metric_label: z.string().optional(),
  strength_value: z.number().optional(),
});

export const RelationsResponseSchema = z.object({
  commit_hash: z.string(),
  relations: z.array(RelationRowSchema),
});

export const ProjectInfoResponseSchema = z.object({
  project_id: z.string().nullable(),
  branch: z.string().nullable(),
  commit_count: z.number(),
  schema_version: z.number(),
  store_present: z.boolean(),
});

export const CommitRowSchema = z.object({
  commit_hash: z.string(),
  commit_order: z.number(),
  authored_at: z.string().nullable(),
  summary: z.string().nullable(),
});

export const commitsResponseSchema = z.array(CommitRowSchema);

export const GraphNodeSchema = z.object({
  module_name: z.string(),
  total_lines: z.number(),
  metrics: z.record(z.string(), z.number()).optional(),
  line_counts: z.record(z.string(), z.number()),
});

export const GraphEdgeSchema = z.object({
  source: z.string(),
  target: z.string(),
  score: z.number(),
  breakdown: EdgeBreakdownSchema.optional(),
  kinds: readonlyStringRecord.optional(),
  kind_occurrence_count: z.number().optional(),
  commit_hash: z.string().optional(),
});

export const GraphResponseSchema = z.object({
  commit_hash: z.string(),
  nodes: z.array(GraphNodeSchema),
  edges: z.array(GraphEdgeSchema),
});

export const TimeseriesPointSchema = z.object({
  commit_order: z.number(),
  commit_hash: z.string(),
  value: z.number().nullable(),
});

export const TimeseriesSeriesSchema = z.object({
  name: z.string(),
  points: z.array(TimeseriesPointSchema),
});

export const TimeseriesResponseSchema = z.object({
  level: z.enum(["module", "file"]),
  metric_id: z.string(),
  agg: z.string(),
  series: z.array(TimeseriesSeriesSchema),
});

export const HotspotItemSchema = z.object({
  name: z.string(),
  current: z.number(),
  first: z.number().nullable().optional(),
  growth: z.number().nullable().optional(),
});

export const HotspotsResponseSchema = z.object({
  by: z.enum(["value", "growth"]),
  items: z.array(HotspotItemSchema),
});

/** RPC response envelope (webview bridge). Discriminated on `status`:
 * exactly one of `result`/`error`, never both. */
export const ResponseEnvelopeSchema = z.discriminatedUnion("status", [
  z.object({
    kind: z.literal("response"),
    status: z.literal("ok"),
    id: z.number(),
    result: z.unknown(),
  }),
  z.object({
    kind: z.literal("response"),
    status: z.literal("error"),
    id: z.number(),
    error: z.object({ code: z.string(), message: z.string() }),
  }),
]);

File: src/ppi/query/schemas.py
"""Transport-neutral API response schemas shared by the FastAPI server and the stdio RPC servant."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel


class CommitResponse(BaseModel):
    commit_hash: str
    commit_order: int
    authored_at: Any | None
    summary: str | None


class TimeseriesPointResponse(BaseModel):
    commit_order: int
    commit_hash: str
    value: float | int | None


class TimeseriesSeriesResponse(BaseModel):
    name: str
    points: list[TimeseriesPointResponse]


class TimeseriesResponse(BaseModel):
    level: Literal["module", "file"]
    metric_id: str
    agg: str
    series: list[TimeseriesSeriesResponse]


class HotspotItemResponse(BaseModel):
    name: str
    current: float
    first: float | None = None
    growth: float | None = None


class HotspotsResponse(BaseModel):
    by: Literal["value", "growth"]
    items: list[HotspotItemResponse]


class EdgeResponse(BaseModel):
    source: str
    target: str
    score: int
    kinds: dict[str, int]
    kind_occurrence_count: int = 0
    breakdown: dict[str, int] | None = None
    commit_hash: str


class GraphNodeResponse(BaseModel):
    module_name: str
    total_lines: int
    metrics: dict[str, float]
    line_counts: dict[str, int]
    line_categories: dict[str, int]


class GraphResponse(BaseModel):
    commit_hash: str
    nodes: list[GraphNodeResponse]
    edges: list[EdgeResponse]


class UiOption(BaseModel):
    id: str
    label: str
    default_enabled: bool = False


class UiMetricOption(BaseModel):
    id: str
    label: str
    unit: str = ""
    format: str = ""
    default_enabled: bool = False


class UiColumnDefinition(BaseModel):
    key: str
    label: str
    type: str = "string"
    metric_id: str | None = None
    width: int | None = None


class UiTableDefinition(BaseModel):
    key: str
    label: str
    columns: list[UiColumnDefinition]


class UiGraphConfig(BaseModel):
    edge_types: list[UiOption]
    line_categories: list[UiOption]
    brightness_metrics: list[UiMetricOption]
    node_size_metrics: list[UiMetricOption]
    link_thickness_metrics: list[UiMetricOption]


class UiConfigResponse(BaseModel):
    dashboard_metrics: list[UiMetricOption]
    aggregations: list[UiOption]
    tables: list[UiTableDefinition]
    graph: UiGraphConfig


class GenericTableRow(BaseModel):
    cells: dict[str, Any]


class GenericTableResponse(BaseModel):
    commit_hash: str
    rows: list[GenericTableRow]


class RelationRowResponse(BaseModel):
    source_id: str
    source_label: str
    target_id: str
    target_label: str
    relation_type_id: str
    relation_type_label: str
    strength_metric_id: str = ""
    strength_metric_label: str = ""
    strength_value: float = 0


class RelationsResponse(BaseModel):
    commit_hash: str
    relations: list[RelationRowResponse]


class ProjectInfoResponse(BaseModel):
    project_id: str | None
    branch: str | None
    commit_count: int
    schema_version: int
    store_present: bool
